python/astro_metadata_translator/translator.py

```diff
@@ -943,6 +946,61 @@ def to_observation_counter(self):
         """
         return 0
 
+    @classmethod
+    def determine_translatable_headers(cls, filename, primary=None):
+        """Given a file return all the headers usable for metadata translation.
+
+        This method can optionally be given a header from the file.  This
+        header will generally be the primary header or a merge of the first
+        two headers.  In the base class implementation it is assumed that
+        this supplied header is the only useful header for metadata translation
+        and it will be returned unchanged if given. This can avoid
+        unnecesarily re-opening the file and re-reading the header when the
+        content is already known.
+
+        If no header is supplied, a header will be read from the supplied
+        file using `read_basic_metadata_from_file`, allowing it to merge
+        the primary and secondary header of a multi-extension FITS file.
+
+        Subclasses can return multiple headers and ignore the externally
+        supplied header.
+
+        Parameters
+        ----------
+        filename : `str`
+            Path to a file in a format understood by this translator.
+        primary : `dict`-like, optional
+            The primary header obtained by the caller. This is sometimes
+            already known, for example if a system is trying to bootstrap
+            without already knowing what data is in the file. For many
+            instruments where the primary header is the only relevant
+            header, the primary header will be returned with no further
+            action.
+
+        Yields
+        ------
+        headers : iterator of `dict`-like
+            Each relevant header in turn. Can be the supplied primary header.
```

Comment:
"Can include"?

I wonder if this is a little too generic, since it makes it hard for the caller to know if the primary header is redundant with the return value. Does this also need to be special-cased by instrument?

Reply:
I'll try to clarify. The method is only required to return a usable header for translation. How the method implements that is up to the method. If it wants to use the supplied header it can or else it can ignore it completely. I was trying not to repeat the general text from above.

Reply:
I wasn't asking about whether it uses the supplied argument, but about whether `headers` includes the primary header at all (including a primary header it's re-read). From the DECam and Megacam examples it looks like you only want to return the primary header if there are no others, and if there are secondary headers then you never want to process the primary header separately?

Reply:
I'm not understanding why the caller needs to know. Some instruments may return some information from the supplied header, other instruments may not. The only reason primary exists is to help out the general case by allowing the file open to be bypassed for instruments where the primary header is going to be enough for translation. MegaPrime will never look at the supplied header because by definition it's irrelevant. DECam will use it if given and merge the detector headers over the top. HSC can return the header immediately because it knows that there is only one header in an HSC file.

Reply:
If primary=None the method still has to return the correct answer.

Reply:
Perhaps say "Includes the supplied primary header, if this header is needed for translation."? That makes it clear that the caller can just trust the output, instead of making the presence/absence of primary sound like an arbitrary choice.

Reply:
The entire method assumes that each header returned by it should be directly usable as input to ObservationInfo constructor.  It shouldn't be trying to check if what was given as the header is identical to what is returned. Presence or absence of primary in returned header is instrument-specific and not something that caller needs to care about.

Reply:
Yes, exactly! But "Can be the supplied primary header." implies otherwise.

Reply:
Ah 😄 -- I'll remove that comment.

Reply:
I had already removed that comment.